### PR TITLE
cmake builds with quiche/quic support

### PR DESCRIPTION
**Pull request by bryancall**

### CODE REVIEW

**Comment by bryancall** [comment]
[approve ci]

**Comment by bryancall** [comment]
[approve ci]

File: cmake/FindQUICHE.cmake [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+INCLUDE(FindPackageHandleStandardArgs)
+FIND_PACKAGE_HANDLE_STANDARD_ARGS(QUICHE DEFAULT_MSG QUICHE_LIBRARY QUICHE_INCLUDE_DIR)
+
+if(QUICHE_FOUND)
+  set(QUICHE_LIBRARY_DIR ${QUICHE_INSTALL_DIR}/lib)
+  set(QUIC_LIBRARY_DIRS ${QUICHE_LIBRARY_DIR})
+  set(QUIC_LIBRARIES ${QUICHE_LIBRARY} http3 quic)
+  set(QUIC_INCLUDE_DIRS ${QUICHE_INCLUDE_DIR} ${CMAKE_SOURCE_DIR}/iocore/net/quic)
+  set(TS_HAS_QUICHE 1)
+  set(TS_USE_QUIC 1)
```
Comment (maskit) [inline cmake/FindQUICHE.cmake:32]: TS_USE_QUIC becomes 1 if the both Quiche and BoringSSL are available.
```
enable_quic=no
TS_CHECK_QUICHE
if test "${has_quiche}" = "1"; then
if test "$openssl_is_boringssl" = "1" ; then
enable_quic=yes
else
  AC_MSG_ERROR([Use of BoringSSL is required if Quiche is used.])
fi
fi
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -60,14 +60,24 @@ include(CheckIncludeFiles)
...
 CHECK_INCLUDE_FILE(linux/hdreg.h HAVE_LINUX_HDREG_H)
 CHECK_INCLUDE_FILE(linux/major.h HAVE_LINUX_MAJOR_H)
+CHECK_INCLUDE_FILE(netdb.h HAVE_NETDB_H)
+CHECK_INCLUDE_FILE(netinet/in.h HAVE_NETINET_IN_H)
+CHECK_INCLUDE_FILE(netinet/in_systm.h HAVE_NETINET_IN_SYSTM_H)
+CHECK_INCLUDE_FILE(netinet/tcp.h HAVE_NETINET_TCP_H)
+CHECK_INCLUDE_FILE(netinet/ip_icmp.h HAVE_NETINET_IP_ICMP_H)
+CHECK_INCLUDE_FILE(siginfo.h HAVE_SIGINFO_H)
+CHECK_INCLUDE_FILE(stdlib.h HAVE_STDLIB_H)
+CHECK_INCLUDE_FILE(stdint.h HAVE_STDINT_H)
```
Comment (JosiahWI) [inline CMakeLists.txt:79]:
```suggestion
CHECK_INCLUDE_FILE(stdint.h HAVE_STDINT_H)
CHECK_INCLUDE_FILE(stdlib.h HAVE_STDLIB_H)
```

File: cmake/FindQUICHE.cmake [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+#  is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing permissions and limitations under
+#  the License.
+#
+#######################
+
+set(QUICHE_INSTALL_DIR /opt/quiche)
+
+find_path(QUICHE_INCLUDE_DIR NAMES quiche.h HINTS ${QUICHE_INSTALL_DIR} PATH_SUFFIXES include)
+find_library(QUICHE_LIBRARY NAMES quiche HINTS ${QUICHE_INSTALL_DIR} PATH_SUFFIXES lib)
```
Comment (JosiahWI) [inline cmake/FindQUICHE.cmake:21]:
```suggestion
find_path(QUICHE_INCLUDE_DIR NAMES quiche.h PATHS ${QUICHE_INSTALL_DIR} PATH_SUFFIXES include)
find_library(QUICHE_LIBRARY NAMES quiche PATHS ${QUICHE_INSTALL_DIR} PATH_SUFFIXES lib)
```

> Search the paths specified by the HINTS option. These should be paths computed by system introspection, such as a hint provided by the location of another item already found. Hard-coded guesses should be specified with the PATHS option.
(https://cmake.org/cmake/help/latest/command/find_library.html)

File: cmake/FindQUICHE.cmake [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+INCLUDE(FindPackageHandleStandardArgs)
```
Comment (JosiahWI) [inline cmake/FindQUICHE.cmake:23]:
```suggestion
include(FindPackageHandleStandardArgs)
```

File: iocore/net/quic/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  QUICStreamManager.cc
+  QUICStreamManager_quiche.cc
+  QUICStreamAdapter.cc
+  QUICStreamVCAdapter.cc
+)
+
+target_include_directories(quic PRIVATE
+  ${CMAKE_SOURCE_DIR}/iocore/net
+  ${CMAKE_SOURCE_DIR}/iocore/net/quic
+  ${CMAKE_SOURCE_DIR}/iocore/eventsystem
```
Comment (JosiahWI) [inline iocore/net/quic/CMakeLists.txt:38]:
```suggestion
  "${CMAKE_SOURCE_DIR}/iocore/net"
  "${CMAKE_SOURCE_DIR}/iocore/net/quic"
  "${CMAKE_SOURCE_DIR}/iocore/eventsystem"
```

File: proxy/http3/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        Http3HeaderVIOAdaptor.cc
+        Http3StreamDataVIOAdaptor.cc
+        QPACK.cc
+)
+
+target_include_directories(http3 PRIVATE
+  ${CMAKE_SOURCE_DIR}/iocore/net
+  ${CMAKE_SOURCE_DIR}/iocore/net/quic
+  ${CMAKE_SOURCE_DIR}/iocore/eventsystem
+  ${CMAKE_SOURCE_DIR}/mgmt
```
Comment (JosiahWI) [inline proxy/http3/CMakeLists.txt:43]:
```suggestion
  "${CMAKE_SOURCE_DIR}/iocore/net"
  "${CMAKE_SOURCE_DIR}/iocore/net/quic"
  "${CMAKE_SOURCE_DIR}/iocore/eventsystem"
  "${CMAKE_SOURCE_DIR}/mgmt"
```

File: cmake/FindQUICHE.cmake [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+FIND_PACKAGE_HANDLE_STANDARD_ARGS(QUICHE DEFAULT_MSG QUICHE_LIBRARY QUICHE_INCLUDE_DIR)
+
+if(QUICHE_FOUND)
+  set(QUICHE_LIBRARY_DIR ${QUICHE_INSTALL_DIR}/lib)
+  set(QUIC_LIBRARY_DIRS ${QUICHE_LIBRARY_DIR})
+  set(QUIC_LIBRARIES ${QUICHE_LIBRARY} http3 quic)
+  set(QUIC_INCLUDE_DIRS ${QUICHE_INCLUDE_DIR} ${CMAKE_SOURCE_DIR}/iocore/net/quic)
```
Comment (JosiahWI) [inline cmake/FindQUICHE.cmake:30]:
```suggestion
  set(QUICHE_LIBRARY_DIR "${QUICHE_INSTALL_DIR}/lib")
  set(QUIC_LIBRARY_DIRS "${QUICHE_LIBRARY_DIR}")
  set(QUIC_LIBRARIES "${QUICHE_LIBRARY}" http3 quic)
  set(QUIC_INCLUDE_DIRS "${QUICHE_INCLUDE_DIR} ${CMAKE_SOURCE_DIR}/iocore/net/quic")
```

**Comment by JosiahWI** [comment]
Indentation seems very inconsistent. I think we're using 4-space indents in all the CMake stuff - CMake itself uses 2-space indents, which I like, but it might be good to have a unified style.

File: cmake/FindQUICHE.cmake [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+#
+#######################
+
+set(QUICHE_INSTALL_DIR /opt/quiche)
+
+find_path(QUICHE_INCLUDE_DIR NAMES quiche.h HINTS ${QUICHE_INSTALL_DIR} PATH_SUFFIXES include)
+find_library(QUICHE_LIBRARY NAMES quiche HINTS ${QUICHE_INSTALL_DIR} PATH_SUFFIXES lib)
+
+INCLUDE(FindPackageHandleStandardArgs)
+FIND_PACKAGE_HANDLE_STANDARD_ARGS(QUICHE DEFAULT_MSG QUICHE_LIBRARY QUICHE_INCLUDE_DIR)
```
Comment (JosiahWI) [inline cmake/FindQUICHE.cmake:24]:
```suggestion
find_package_handle_standard_args(QUICHE DEFAULT_MSG QUICHE_LIBRARY QUICHE_INCLUDE_DIR)
```

**Comment by bryancall** [comment]
#9892 superseded this PR

**Pull request closed without merging**